File: tests/org/elixir_lang/reference/callable/IntramoduleTest.java
package org.elixir_lang.reference.callable;

import com.intellij.navigation.ItemPresentation;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Disposer;
import com.intellij.psi.*;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import org.elixir_lang.psi.ElixirIdentifier;

import static org.junit.Assert.assertNotEquals;

public class IntramoduleTest extends BasePlatformTestCase {
    /*
     * Tests
     */

    public void testAmbiguousBackReference() {
        myFixture.configureByFiles("ambiguous_back.ex");
        PsiElement ambiguous = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getPrevSibling();

        assertInstanceOf(ambiguous.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = ambiguous.getReference();

        assertNotNull("`referenced` has no reference", reference);

        PsiElement resolved = reference.resolve();

        assertNotNull("`referenced` not resolved", resolved);
        assertEquals(
                "ambiguous reference does not resolve to previous function declaration",
                "def referenced do\n\n  end",
                resolved.getText()
        );
    }

    public void testAmbiguousForwardReference() {
        myFixture.configureByFiles("ambiguous_forward.ex");
        PsiElement ambiguous = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getPrevSibling();

        assertInstanceOf(ambiguous.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = ambiguous.getReference();

        assertNotNull("`referenced` has no reference", reference);

        PsiElement resolved = reference.resolve();

        assertNotNull("`referenced` not resolved", resolved);
        assertEquals(
                "ambiguous reference does not resolve to forward function declaration",
                "def referenced do\n\n  end",
                resolved.getText()
        );
    }

    public void testAmbiguousRecursiveReference() {
        myFixture.configureByFiles("ambiguous_recursive.ex");
        PsiElement ambiguous = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getPrevSibling();

        assertInstanceOf(ambiguous.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = ambiguous.getReference();

        assertNotNull("`referenced` has no reference", reference);

        PsiElement resolved = reference.resolve();

        assertNotNull("`referenced` not resolved", resolved);
        assertEquals(
                "ambiguous reference does not resolve to recursive function declaration",
                "def referenced do\n    referenced\n\n    a = 1\n  end",
                resolved.getText()
        );
    }

    public void testFunctionNameMultipleSameArity() {
        myFixture.configureByFiles("function_name_multiple_same_arity.ex");
        PsiElement parenthesesCall = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getParent()
                .getParent()
                .getParent();

        assertInstanceOf(parenthesesCall.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = parenthesesCall.getReference();

        assertNotNull("`referenced` has no reference", reference);

        assertInstanceOf(reference, PsiPolyVariantReference.class);

        PsiPolyVariantReference polyVariantReference = (PsiPolyVariantReference) reference;

        ResolveResult[] resolveResults = polyVariantReference.multiResolve(false);

        assertNotEquals("Resolved to both clauses instead of selected clause", 2, resolveResults.length);
        assertEquals("Resolves to self", 1, resolveResults.length);

        PsiElement resolved = reference.resolve();

        assertNotNull("Reference not resolved", resolved);
        assertEquals("def referenced(true) do\n  end", resolved.getText());
    }

    public void testParenthesesRecursiveReference() {
        myFixture.configureByFiles("parentheses_recursive.ex");
        PsiElement parenthesesCall = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getParent()
                .getParent()
                .getParent();

        assertInstanceOf(parenthesesCall.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = parenthesesCall.getReference();

        assertNotNull("`referenced` has no reference", reference);

        PsiElement resolved = reference.resolve();

        assertNotNull("`referenced` not resolved", resolved);
        assertEquals(
                "ambiguous reference does not resolve to recursive function declaration",
                "def referenced do\n    referenced()\n\n    a = 1\n  end",
                resolved.getText()
        );
    }

    public void testParenthesesMultipleCorrectArityReference() {
        myFixture.configureByFiles("parentheses_multiple_correct_arity.ex");
        PsiElement parenthesesCall = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getParent()
                .getParent()
                .getParent();

        assertInstanceOf(parenthesesCall.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = parenthesesCall.getReference();

        assertNotNull("`referenced` has no reference", reference);

        assertInstanceOf(reference, PsiPolyVariantReference.class);

        PsiPolyVariantReference polyVariantReference = (PsiPolyVariantReference) reference;

        ResolveResult[] resolveResults = polyVariantReference.multiResolve(false);

        assertEquals("Did not resolve to all clauses", 2, resolveResults.length);

        ResolveResult firstResolveResult = resolveResults[0];
        PsiElement firstResolved = firstResolveResult.getElement();

        assertEquals(
                "first ResolveResult is not the true clause",
                "def referenced(true) do\n  end",
                firstResolved.getText()
        );

        assertInstanceOf(firstResolved, NavigatablePsiElement.class);

        NavigatablePsiElement navigatablefirstResolved = (NavigatablePsiElement) firstResolved;

        ItemPresentation firstPresentation = navigatablefirstResolved.getPresentation();

        assertNotNull("first ResolveResult element has no presentation", firstPresentation);

        assertEquals("/src/parentheses_multiple_correct_arity.ex defmodule A", firstPresentation.getLocationString());
        assertEquals("referenced(true)", firstPresentation.getPresentableText());

        ResolveResult secondResolveResult = resolveResults[1];
        PsiElement secondResolved = secondResolveResult.getElement();

        assertEquals(
                "second ResolveResult is not the false clause",
                "def referenced(false) do\n  end",
                secondResolved.getText()
        );

        assertInstanceOf(secondResolved, NavigatablePsiElement.class);

        NavigatablePsiElement navigatablesecondResolved = (NavigatablePsiElement) secondResolved;

        ItemPresentation secondPresentation = navigatablesecondResolved.getPresentation();

        assertNotNull("second ResolveResult element has no presentation", secondPresentation);

        assertEquals("/src/parentheses_multiple_correct_arity.ex defmodule A", secondPresentation.getLocationString());
        assertEquals("referenced(false)", secondPresentation.getPresentableText());
    }

    public void testParenthesesSingleCorrectArityReference() {
        myFixture.configureByFiles("parentheses_single_correct_arity.ex");
        PsiElement parenthesesCall = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset())
                .getParent()
                .getParent()
                .getParent();

        assertInstanceOf(parenthesesCall.getFirstChild(), ElixirIdentifier.class);

        PsiReference reference = parenthesesCall.getReference();

        assertNotNull("`referenced` has no reference", reference);

        assertInstanceOf(reference, PsiPolyVariantReference.class);

        PsiPolyVariantReference polyVariantReference = (PsiPolyVariantReference) reference;

        ResolveResult[] resolveResults = polyVariantReference.multiResolve(false);

        assertEquals("Did not resolve to all clauses", 1, resolveResults.length);

        ResolveResult firstResolveResult = resolveResults[0];

        assertTrue("first ResolveResult is not a valid result", firstResolveResult.isValidResult());

        PsiElement firstResolved = firstResolveResult.getElement();

        assertEquals(
                "first ResolveResult is not 1-arity clause",
                "def referenced(_) do\n" +
                        "  end",
                firstResolved.getText()
        );
    }

    /*
     * Protected Instance Methods
     */

    @Override
    protected String getTestDataPath() {
        return "testData/org/elixir_lang/reference/callable/intramodule";
    }


    @Override
    protected void tearDown() throws Exception {
        super.tearDown();

        if (myFixture != null) {
            Project project = getProject();

            if (project != null && !project.isDisposed()) {
                Disposer.dispose(project);
            }
        }
    }
}

File: tests/org/elixir_lang/reference/callable/IntramoduleTest.kt
package org.elixir_lang.reference.callable

import com.intellij.openapi.util.Disposer
import org.elixir_lang.psi.ElixirIdentifier
import com.intellij.psi.PsiPolyVariantReference
import com.intellij.psi.NavigatablePsiElement
import org.elixir_lang.PlatformTestCase
import org.junit.Assert
import java.lang.Exception

class IntramoduleTest : PlatformTestCase() {
    fun testAmbiguousBackReference() {
        myFixture.configureByFiles("ambiguous_back.ex")
        val ambiguous = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .prevSibling
        assertInstanceOf(ambiguous.firstChild, ElixirIdentifier::class.java)

        val reference = ambiguous.reference
        assertNotNull("`referenced` has no reference", reference)

        val resolved = reference!!.resolve()
        assertNotNull("`referenced` not resolved", resolved)
        assertEquals(
                "ambiguous reference does not resolve to previous function declaration",
                "def referenced do\n\n  end",
                resolved!!.text
        )
    }

    fun testAmbiguousForwardReference() {
        myFixture.configureByFiles("ambiguous_forward.ex")
        val ambiguous = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .prevSibling
        assertInstanceOf(ambiguous.firstChild, ElixirIdentifier::class.java)

        val reference = ambiguous.reference
        assertNotNull("`referenced` has no reference", reference)

        val resolved = reference!!.resolve()
        assertNotNull("`referenced` not resolved", resolved)
        assertEquals(
                "ambiguous reference does not resolve to forward function declaration",
                "def referenced do\n\n  end",
                resolved!!.text
        )
    }

    fun testAmbiguousRecursiveReference() {
        myFixture.configureByFiles("ambiguous_recursive.ex")
        val ambiguous = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .prevSibling
        assertInstanceOf(ambiguous.firstChild, ElixirIdentifier::class.java)

        val reference = ambiguous.reference
        assertNotNull("`referenced` has no reference", reference)

        val resolved = reference!!.resolve()
        assertNotNull("`referenced` not resolved", resolved)
        assertEquals(
                "ambiguous reference does not resolve to recursive function declaration",
                "def referenced do\n    referenced\n\n    a = 1\n  end",
                resolved!!.text
        )
    }

    fun testFunctionNameMultipleSameArity() {
        myFixture.configureByFiles("function_name_multiple_same_arity.ex")
        val parenthesesCall = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .parent
                .parent
                .parent
        assertInstanceOf(parenthesesCall.firstChild, ElixirIdentifier::class.java)

        val reference = parenthesesCall.reference
        assertNotNull("`referenced` has no reference", reference)
        assertInstanceOf(reference, PsiPolyVariantReference::class.java)
        val polyVariantReference = reference as PsiPolyVariantReference?

        val resolveResults = polyVariantReference!!.multiResolve(false)
        Assert.assertNotEquals("Resolved to both clauses instead of selected clause", 2, resolveResults.size.toLong())
        assertEquals("Resolves to self", 1, resolveResults.size)

        val resolved = reference!!.resolve()
        assertNotNull("Reference not resolved", resolved)
        assertEquals("def referenced(true) do\n  end", resolved!!.text)
    }

    fun testParenthesesRecursiveReference() {
        myFixture.configureByFiles("parentheses_recursive.ex")
        val parenthesesCall = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .parent
                .parent
                .parent
        assertInstanceOf(parenthesesCall.firstChild, ElixirIdentifier::class.java)

        val reference = parenthesesCall.reference
        assertNotNull("`referenced` has no reference", reference)

        val resolved = reference!!.resolve()
        assertNotNull("`referenced` not resolved", resolved)
        assertEquals(
                "ambiguous reference does not resolve to recursive function declaration",
                "def referenced do\n    referenced()\n\n    a = 1\n  end",
                resolved!!.text
        )
    }

    fun testParenthesesMultipleCorrectArityReference() {
        myFixture.configureByFiles("parentheses_multiple_correct_arity.ex")
        val parenthesesCall = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .parent
                .parent
                .parent
        assertInstanceOf(parenthesesCall.firstChild, ElixirIdentifier::class.java)

        val reference = parenthesesCall.reference
        assertNotNull("`referenced` has no reference", reference)
        assertInstanceOf(reference, PsiPolyVariantReference::class.java)
        val polyVariantReference = reference as PsiPolyVariantReference?

        val resolveResults = polyVariantReference!!.multiResolve(false)
        assertEquals("Did not resolve to all clauses", 2, resolveResults.size)

        val firstResolveResult = resolveResults[0]
        val firstResolved = firstResolveResult.element
        assertEquals(
                "first ResolveResult is not the true clause",
                "def referenced(true) do\n  end",
                firstResolved!!.text
        )
        assertInstanceOf(firstResolved, NavigatablePsiElement::class.java)
        val navigatablefirstResolved = firstResolved as NavigatablePsiElement?
        val firstPresentation = navigatablefirstResolved!!.presentation
        assertNotNull("first ResolveResult element has no presentation", firstPresentation)
        assertEquals("/src/parentheses_multiple_correct_arity.ex defmodule A", firstPresentation!!.locationString)
        assertEquals("referenced(true)", firstPresentation.presentableText)

        val secondResolveResult = resolveResults[1]
        val secondResolved = secondResolveResult.element
        assertEquals(
                "second ResolveResult is not the false clause",
                "def referenced(false) do\n  end",
                secondResolved!!.text
        )
        assertInstanceOf(secondResolved, NavigatablePsiElement::class.java)
        val navigatablesecondResolved = secondResolved as NavigatablePsiElement?
        val secondPresentation = navigatablesecondResolved!!.presentation
        assertNotNull("second ResolveResult element has no presentation", secondPresentation)
        assertEquals("/src/parentheses_multiple_correct_arity.ex defmodule A", secondPresentation!!.locationString)
        assertEquals("referenced(false)", secondPresentation.presentableText)
    }

    fun testParenthesesSingleCorrectArityReference() {
        myFixture.configureByFiles("parentheses_single_correct_arity.ex")
        val parenthesesCall = myFixture
                .file
                .findElementAt(myFixture.caretOffset)!!
                .parent
                .parent
                .parent
        assertInstanceOf(parenthesesCall.firstChild, ElixirIdentifier::class.java)

        val reference = parenthesesCall.reference
        assertNotNull("`referenced` has no reference", reference)
        assertInstanceOf(reference, PsiPolyVariantReference::class.java)
        val polyVariantReference = reference as PsiPolyVariantReference

        val resolveResults = polyVariantReference.multiResolve(false)
        assertEquals("Did not resolve to all clauses", 1, resolveResults.size)

        val firstResolveResult = resolveResults[0]
        assertTrue("first ResolveResult is not a valid result", firstResolveResult.isValidResult)
        val firstResolved = firstResolveResult.element
        assertEquals(
                "first ResolveResult is not 1-arity clause",
                """def referenced(_) do
  end""",
                firstResolved!!.text
        )
    }

    override fun getTestDataPath(): String = "testData/org/elixir_lang/reference/callable/intramodule"

    @Throws(Exception::class)
    override fun tearDown() {
        super.tearDown()
        if (myFixture != null) {
            val project = project
            if (project != null && !project.isDisposed) {
                Disposer.dispose(project)
            }
        }
    }
}
